test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,172 @@
+using grpc::ServerContext;
+using grpc::Status;
+
+#include <gtest/gtest.h>
+
+// Test variables
+std::string server_address("0.0.0.0:10000");
+std::string custom_credentials_type("INSECURE_CREDENTIALS");
+std::string sampling_times = "2";
+std::string sampling_interval_seconds = "5";
+std::string output_json("./output.json");
+
+// Creata an echo server - randomly delay 0.1 to 0.2 s
+class EchoServerImpl final : public grpc::testing::TestService::Service {
+  Status EmptyCall(::grpc::ServerContext* context,
+                   const grpc::testing::Empty* request,
+                   grpc::testing::Empty* response) {
+    srand(unsigned(time(0)));
+    unsigned int server_delay_microseconds = 100000;
+    server_delay_microseconds += rand() % server_delay_microseconds;
+    usleep(server_delay_microseconds);
+    return Status::OK;
+  }
+};
+
+// Run server in a thread
+void RunServer() {
+  // register channelz service
+  ::grpc::channelz::experimental::InitChannelzService();
+
+  EchoServerImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc_impl::reflection::InitProtoReflectionServerBuilderPlugin();
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  server->Wait();
+}
+
+// Creata an echo client - set timeout as 0.15s
+class EchoClientImpl {
+ public:
+  EchoClientImpl(std::shared_ptr<Channel> channel)
+      : stub_(grpc::testing::TestService::NewStub(channel)) {}
+  Status EmptyCall() {
+    grpc::testing::Empty request;
+    grpc::testing::Empty response;
+    ClientContext context;
+    int64_t timeout_microseconds = 150;
+    context.set_deadline(
+        grpc_timeout_milliseconds_to_deadline(timeout_microseconds));
+    Status status = stub_->EmptyCall(&context, request, &response);
+    return status;
+  }
+
+ private:
+  std::unique_ptr<grpc::testing::TestService::Stub> stub_;
+};
+
+// Run client in a thread
+void RunClient(std::string client_id) {
+  // std::string target_str = "localhost:10000";
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  EchoClientImpl echoer(grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
```

Comment:
we'll need to use some signal to know when to terminate the loop upon cleanup/thread-join

Suggest using `gpr_event`. See e.g. https://github.com/grpc/grpc/blob/8febefc05784fe2e12a5933a96d373297ad0cfdf/test/cpp/naming/resolver_component_test.cc#L312

Reply:
Done